src/js/common/parser.js

```diff
@@ -0,0 +1,223 @@
+/*
```

Comment:
Eventually we will have USER TOOLS. That is, in response to some diagnostic from the system, they should be guided, visually, towards an explanation of what caused it and what they can do to make it go away. The end users may or may not be aware "that we're using JSON Schemas at all" but hopefully they will be aware that we have built a system that does more than just push diagnostics at them in the form of textual messages. Whilst this day is some way off, we should do our best to consciously build scaffolding for it.